filetracker/servers/storage.py

```diff
@@ -108,62 +109,56 @@ def store(self, name, data, version, size=0, compressed=False, digest=None):
-                        digest = file_digest(compressed_file)
+                    # This shouldn't occur if the request came from a proper
+                    # filetracker client, so we don't care if it's slow.
+                    with gzip.open(
+                            contents.current_path, 'rb') as decompressed:
+                        digest = file_digest(decompressed)
+                    with gzip.open(
+                            contents.current_path, 'rb') as decompressed:
+                        logical_size = _read_stream_for_size(decompressed)
                 else:
-                    digest = file_digest(temp_file_path)
+                    digest = file_digest(contents.current_path)
+                    logical_size = os.stat(contents.current_path).st_size
 
             blob_path = self._blob_path(digest)
             
             with self._lock_blob_with_txn(digest) as txn:
-                digest_bytes = digest.encode('utf8')
-                try:
-                    link_count = int(self.db.get(digest_bytes, 0, txn=txn))
-                except KeyError:
-                    link_count = 0
+                digest_bytes = digest.encode()
 
-                new_count = str(link_count + 1).encode('utf8')
+                link_count = int(self.db.get(digest_bytes, 0, txn=txn))
+                new_count = str(link_count + 1).encode()
                 self.db.put(digest_bytes, new_count, txn=txn)
 
+                # Create a new blob if this isn't a duplicate.
                 if link_count == 0:
-                    # Create a new blob.
                     _create_file_dirs(blob_path)
+                    self.db.put('{}:logical_size'.format(digest).encode(),
+                                str(logical_size).encode())
+
                     if compressed:
-                        if temp_file_path:
-                            shutil.move(temp_file_path, blob_path)
-                        else:
-                            with open(blob_path, 'wb') as blob:
-                                _copy_stream(data, blob, size)
+                        contents.save(blob_path)
                     else:
-                        if temp_file_path:
-                            with open(temp_file_path, 'rb') as raw,\
-                                    gzip.open(blob_path, 'wb') as blob:
-                                shutil.copyfileobj(raw, blob)
-                        else:
-                            with gzip.open(blob_path, 'wb') as blob:
-                                _copy_stream(data, blob, size)
+                        contents.save()
+                        with open(contents.current_path, 'rb') as raw,\
+                                gzip.open(blob_path, 'wb') as blob:
+                            shutil.copyfileobj(raw, blob)
 
-                if temp_file_path and os.path.exists(temp_file_path):
-                    os.unlink(temp_file_path)
+                contents.cleanup()
```

Comment:
I'm slightly worried this may not be executed if something throws an exception (however unlikely should that be).
I guess it's not super important, because it's just leaving a temporary file, but maybe adding a try + finally may be a good idea? What do you think?